gorm/transaction_test.go

```diff
@@ -271,80 +301,148 @@ func TestContext(t *testing.T) {
 }
 
 func TestBeginFromContext_Good(t *testing.T) {
-	ctx := context.Background()
-	db, mock, err := sqlmock.New()
-	if err != nil {
-		t.Fatalf("failed to create sqlmock - %s", err)
-	}
-	gdb, err := gorm.Open("postgres", db)
-	if err != nil {
-		t.Fatalf("failed to open gorm db - %s", err)
-	}
-	mock.ExpectBegin()
-
-	// Case: All good
-	ctxtxn := &Transaction{parent: gdb}
-	ctx = NewContext(ctx, ctxtxn)
-
-	txn1, err := BeginFromContext(ctx)
-	if txn1 == nil {
-		t.Error("Did not receive a transaction from context")
-	}
-	if err != nil {
-		t.Error("Received an error beginning transaction")
-	}
-	if err = mock.ExpectationsWereMet(); err != nil {
-		t.Errorf("failed to begin transaction - %s", err)
-	}
-
-	// Case: Transaction begin is idempotent
-	txn2, err := BeginFromContext(ctx)
-	if txn2 != txn1 {
-		t.Error("Got a different txn than was opened before")
-	}
-	if err != nil {
-		t.Error("Received an error opening transaction")
+	tests := []struct {
+		desc     string
+		withOpts bool
+	}{
+		{
+			desc:     "begin without options",
+			withOpts: false,
+		},
+		{
+			desc:     "begin with options",
+			withOpts: true,
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			ctx := context.Background()
+			opt := &sql.TxOptions{Isolation: sql.LevelSerializable}
+
+			db, mock, err := sqlmock.New()
+			if err != nil {
+				t.Fatalf("failed to create sqlmock - %s", err)
+			}
+			gdb, err := gorm.Open("postgres", db)
+			if err != nil {
+				t.Fatalf("failed to open gorm db - %s", err)
+			}
+			mock.ExpectBegin()
+
+			// Case: All good
+			ctxtxn := &Transaction{parent: gdb}
+			ctx = NewContext(ctx, ctxtxn)
+
+			var txn1, txn2 *gorm.DB
+			switch test.withOpts {
+			case true:
+				txn1, err = BeginWithOptionsFromContext(ctx, opt)
+			case false:
+				txn1, err = BeginFromContext(ctx)
+			}
```

Comment:
I do not really like this `switch` statements happening multiple times. Is it possible to pass function that creates transaction to t.Run instead of having multiple switch blocks?